price-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn deposit(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_active()?;
+
+        let accepted_token_id = self.accepted_token_id().get();
+        require!(payment_token == accepted_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        let caller = self.blockchain().get_caller();
+        self.mint_and_send_redeem_token(&caller, ACCEPTED_TOKEN_REDEEM_NONCE, &payment_amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn withdraw(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_active()?;
+
+        let redeem_token_id = self.redeem_token_id().get();
+        require!(
+            payment_token == redeem_token_id && payment_nonce == ACCEPTED_TOKEN_REDEEM_NONCE,
+            INVALID_PAYMENT_ERR_MSG
+        );
+
+        self.burn_redeem_token(ACCEPTED_TOKEN_REDEEM_NONCE, &payment_amount);
+
+        let caller = self.blockchain().get_caller();
+        let accepted_token_id = self.accepted_token_id().get();
+        self.send()
+            .direct(&caller, &accepted_token_id, 0, &payment_amount, &[]);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn redeem(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_deposit_period_ended()?;
+        require!(!self.lp_token_id().is_empty(), "Pool not created yet");
+
+        let redeem_token_id = self.redeem_token_id().get();
+        require!(payment_token == redeem_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        self.burn_redeem_token(0, &payment_amount);
+
+        let lp_token_amount = self.compute_lp_amount_to_send(payment_nonce, payment_amount);
+        require!(lp_token_amount > 0, "Nothing to redeem");
+
+        let caller = self.blockchain().get_caller();
+        let lp_token_id = self.lp_token_id().get();
+        self.send()
+            .direct(&caller, &lp_token_id, 0, &lp_token_amount, &[]);
+
+        Ok(())
+    }
+
+    // private
+
+    fn require_active(&self) -> SCResult<()> {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let start_epoch = self.start_epoch().get();
+        let end_epoch = self.end_epoch().get();
+        require!(
+            start_epoch <= current_epoch,
+            "Deposit period not started yet"
+        );
+        require!(current_epoch < end_epoch, "Deposit period ended");
+
+        let launched_token_id = self.launched_token_id().get();
+        let current_launched_token_balance =
```

Comment:
no need for this check as there is no need for the initial deposit.

Reply:
Removed.